### PR TITLE
[ENH] ForecastX enable fitting forecaster_y on forecasted X

### PR DESCRIPTION
**Pull request by benHeid**
Enable forecastX to train forecaster_y on prediction of forecaster_X to enable forecaster_y to consider the uncertainty of the forecasted X.

#### Reference Issues/PRs
fixes partially #5225

#### What does this implement/fix? Explain your changes.
This PR adds a parameter to ForecastX that enables to control if ForecastX should fit the internal forecaster_y on the ground truth data of `X` or of the forecasted `X`. Fitting on the forecasted `X` enables `forecaster_y` to account the uncertainty associated with forecast. 

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?

* is there a better name for the newly introduced parameter? 
* should we enable also the fitting on a validation set as proposed in the original issue. Should this be implemented by passing an integer (cutoff) or a float (share of training/validation data) to the newly introduced keyword? 

#### Did you add any tests for the change?
Yes.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/forecasting/compose/_pipeline.py [hunk 1]
```diff
@@ -1199,6 +1199,11 @@ class ForecastX(BaseForecaster):
     columns : None, or pandas compatible index iterator (e.g., list of str), optional
         default = None = all columns in X are used for forecast
         columns to which `forecaster_X` is applied
+    fit_behaviour : str, one of "ues_gt", "ues_forecast", optional,
+        default = "ues_gt"
```
Comment (yarnabrina) [inline sktime/forecasting/compose/_pipeline.py:1203]: I'm not sure if `use_gt` is immediately obvious. I'll suggest `use_actual` or something similar.

File: sktime/forecasting/compose/_pipeline.py [hunk 2]
```diff
@@ -1315,7 +1327,11 @@ def _fit(self, y, X, fh):
             self.forecaster_X_.fit(y=self._get_Xcols(X), fh=fh_X)
 
         self.forecaster_y_ = self.forecaster_y.clone()
-        self.forecaster_y_.fit(y=y, X=X, fh=fh)
+        if self.fit_behaviour == "use_gt":
+            self.forecaster_y_.fit(y=y, X=X, fh=fh)
+        elif self.fit_behaviour == "use_forecast":
+            x_insample = self.forecaster_X_.predict(fh=X.index, X=X)
```
Comment (yarnabrina) [inline sktime/forecasting/compose/_pipeline.py:1333]: Not all of the models support in-sample prediction, there's a boolean tag `capability:insample` for this.

File: sktime/forecasting/compose/tests/test_pipeline.py [hunk 3]
```diff
@@ -478,6 +478,50 @@ def test_forecastx_logic():
     assert np.allclose(y_pred, y_pred_manual)
 
 
+@pytest.mark.skipif(
+    not _check_soft_dependencies("statsmodels", severity="none"),
```
Comment (yarnabrina) [inline sktime/forecasting/compose/tests/test_pipeline.py:482]: Where is `statsmodels` required? If I'm not wrong, `ARIMA` comes from `pmdarima`.

**Review by yarnabrina — COMMENTED** [review]
I'm not sure I understand the motivation of this PR. Can you please explain in what scenario one would like to train their endogenous forecaster using predicted exogenous variables instead of using the actuals for historical data? For future forecasts where X data is unavailable, there's no choice, but why will someone prefer to risk error propagation intentionally when that can be avoided?

Also, the error in in-sample predictions and error in out-of-sample predictions are unlikely to be similarly distributed in my understanding, as the in-sample data has been exposed to the model. Please let me know if I am missing something.

File: sktime/forecasting/compose/_pipeline.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_pipeline.py:1203]: there are typos too. Agree with improving naming convention

File: sktime/forecasting/compose/tests/test_pipeline.py (hunk 3 above)
Comment (fkiraly) [reply sktime/forecasting/compose/tests/test_pipeline.py:482]: yes - `_check_estimator_dep` is safer as it retrieves the correct dependencies.

**Comment by fkiraly** [comment]
> Can you please explain in what scenario one would like to train their endogenous forecaster using predicted exogenous variables instead of using the actuals for historical data?

If I'm not mistaken, it is not a bad idea. In the case of a perfect model, using forecasts result in "denoised" data, i.e., it removes the error term. Given that the same is true for the model in its deployment phase, you might expect more accurate forecasts from any model learning an "`X` - to `y`" prediction rule, as the quality of the `X` in fitting and prediction will not differ.

**Comment by benHeid** [comment]
There is no clear obvious choice. But there are several indications that it might be beneficial to use forecasts in some cases. @fkiraly pointed out a theoretical reason. But I think there are also some practical reasons:
- if you are using actuals you are assuming that your forecasts of x will be accurate or even perfect. However, this is a wrong assumption. The problem here is in the end that the model for y may rely too strong on this assumption and this harms the result. If you are investigating this from a concept drift point of view, you might can say that you have a concept drift right after training,since the probability distribution of a data source (forecast of x vs actuals) is changing, since the forecast of x is not perfect. 
- your forecast of x will probably have a bias. However if you are training on the actuals you will not consider this during training.

As an example you can take a look on the example in the issue.

Hope this helps to understand the intentions behind this PR.

File: sktime/forecasting/compose/_pipeline.py (hunk 2 above)
Comment (benHeid) [reply sktime/forecasting/compose/_pipeline.py:1333]: Thanks. I added a check.

File: sktime/forecasting/compose/tests/test_pipeline.py (hunk 3 above)
Comment (benHeid) [reply sktime/forecasting/compose/tests/test_pipeline.py:482]: I changed it

File: sktime/forecasting/compose/_pipeline.py (hunk 1 above)
Comment (benHeid) [reply sktime/forecasting/compose/_pipeline.py]: use_actual is a good idea.

File: sktime/forecasting/compose/_pipeline.py [hunk 4]
```diff
@@ -1315,7 +1332,16 @@ def _fit(self, y, X, fh):
             self.forecaster_X_.fit(y=self._get_Xcols(X), fh=fh_X)
 
         self.forecaster_y_ = self.forecaster_y.clone()
-        self.forecaster_y_.fit(y=y, X=X, fh=fh)
+        if self.fit_behaviour == "use_actual":
+            self.forecaster_y_.fit(y=y, X=X, fh=fh)
+        elif self.fit_behaviour == "use_forecast":
+            if "capability:insample" not in self.forecaster_X_.get_tags():
```
Comment (yarnabrina) [inline sktime/forecasting/compose/_pipeline.py:1338]: Is it checking if in sample prediction capability is present or not? Or, is it checking if the tag is present? Won't it be present always?

```suggestion
            if not self.forecaster_X_.get_tag("capability:insample"):
```

**Review by yarnabrina — CHANGES_REQUESTED** [review]

File: sktime/forecasting/compose/_pipeline.py (hunk 4 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_pipeline.py:1342]: > Won't it be present always?

Yes, the condition should e the value (True or False), not presence of the tag. This is a logic error.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Logic error that @yarnabrina noticed, otherwise fine.

I also added a commit (I hope you do not mind):

* formats docstring correctly
* removes author credits from the file, these are overall incorrect (e.g., I wrote initial version of ForecastX). We should split the file up and restore them ased on commit history for the individual estimators.

**Review by fkiraly — APPROVED** [review]
